### ISSUE TITLE
`#blank?` fails on non-UTF8 strings

### ISSUE THREAD

**Issue by BrentWheeldon** [rails/rails#28953]
### Steps to reproduce

```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  # Activate the gem you are reporting the issue against.
  gem "activemodel", "5.1.0"
end

require "active_model"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

class BugTest < Minitest::Test
  def test_blankness
    non_utf8_string = "my value".encode("UTF-16LE")

    assert_false non_utf8_string.blank?
  end
end
```

### Expected behavior
`#blank?` should return false if the string is not blank.

### Actual behavior
An `Encoding::CompatibilityError` error is raised

### System configuration
**Rails version**: 5.1.0 (also tested on 5.0.2)

**Ruby version**: 2.4.0

**Comment by alexcameron89** [comment]
Here is a back-trace of the error:
```
Encoding::CompatibilityError: incompatible encoding regexp match (US-ASCII regexp with UTF-16LE string)
   rails/activesupport/lib/active_support/core_ext/object/blank.rb:120:in `match?'
   rails/activesupport/lib/active_support/core_ext/object/blank.rb:120:in `blank?'
   28953.rb:25:in `test_blankness'
```
The error happens when `blank?` attempts a Regex `match?` [here](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/blank.rb#L120).

**Comment by sivagollapalli** [comment]
@alexcameron89 I think encode every string to `UTF-8` before comparision would avoid these kind of errors i.e.,`BLANK_RE.match(self.encode('utf-8'))` . Thoughts please

**Comment by alexcameron89** [comment]
@sivagollapalli doing so can cause issues with certain encodings like ASCII-8BIT.

@fxn since you spent some time on `#blank?` [last year](https://github.com/rails/rails/commit/bbb84a17213d60f25a61ae4a09111d5adac01a05), do you have any thoughts on this?

**Comment by fxn** [comment]
@alexcameron89 I think this is a bug. I'll work on it.

**Comment by fxn** [comment]
The tricky part here is not to make the predicate slower for the main use case.

The logic used by MRI to see if encodings are compatible [involves a few things](https://github.com/ruby/ruby/blob/f033cfa3816e0138906aeaf8adcd3c18a623d16a/re.c#L1380-L1417). We cannot do that by hand.

An optimistic implementation would wrap the call and catch the exception for example, to encode the regexp with `self.encoding` or any other possible way to make them compatible. But `rescue` would slow down all client code for UTF-8, it would be a regression. This is a case of making something fast by making it wrong 😄 .

Let me summon @tenderlove and @matthewd, also @schneems.

**Comment by matthewd** [comment]
Untested (and certainly unprofiled) thinking out loud:

```ruby
ENCODED_BLANKS = Hash.new do |h, enc|
  h[enc] = Regexp.new(BLANK_RE.source.encode(enc), BLANK_RE.options | Regexp::FIXEDENCODING)
end
```

```ruby
def blank?
  # (existing comment that this check has negligible cost and measurable gain)
  empty? ||

    # For simple ASCII strings, we can avoid the allocation from calling self.encoding.
    # That seems likely to be worthwhile -- but should obviously be proven.
    if ascii_only?
      BLANK_RE.match?(self)
    else
      ENCODED_BLANKS[encoding].match?(self)
    end
end
```

**Comment by schneems** [comment]
Would we consider using a gem like fast blank? Not sure if they've got the same problem but it is certainly is faster.

**Comment by BrentWheeldon** [comment]
@schneems FWIW that's what we ended up doing ourselves and it solved this issue.

**Comment by gwincr11** [comment]
I did some benchmarking on the propose solution from @matthewd and a few alternatives. 
```ruby
require 'benchmark'

class String
  BLANK_RE = /\A[[:space:]]*\z/
  ENCODED_BLANKS = Concurrent::Map.new do |h, enc|
      h[enc] = Regexp.new(BLANK_RE.source.encode(enc), BLANK_RE.options | Regexp::FIXEDENCODING)
  end

  # A string is blank if it's empty or contains whitespaces only:
  #
  #   ''.blank?       # => true
  #   '   '.blank?    # => true
  #   "\t\n\r".blank? # => true
  #   ' blah '.blank? # => false
  #
  # Unicode whitespace is supported:
  #
  #   "\u00a0".blank? # => true
  #
  # @return [true, false]
  def blank?
    # The regexp that matches blank strings is expensive. For the case of empty
    # strings we can speed up this method (~3.5x) with an empty? call. The
    # penalty for the rest of strings is marginal.
    n = 50000
    encoding_required = encoding.to_s != "UTF-8"
    puts encoding
    if !encoding_required
      Benchmark.bm do |x|
        x.report("Current code: ") { empty? || BLANK_RE.match?(self) }
      end
    end
    Benchmark.bm do |x|
      x.report("always encode (encode change? #{encoding_required}): ") do
        empty? || blank_re.match?(self)
      end
      x.report("Conditionally encode (encode change? #{encoding_required}): ") do
        empty? ||
          if ascii_only?
            BLANK_RE.match?(self)
          else
            blank_re.match?(self)
          end
      end
      x.report("Try Catch (encode change? #{encoding_required}): ") do
        empty? ||
          begin
            BLANK_RE.match?(self)
          rescue Encoding::CompatibilityError
            blank_re.match?(self)
          end
      end
      x.report("Concurrent map and if  (encode change? #{encoding_required}): ") do
        empty? ||
          if ascii_only?
            BLANK_RE.match?(self)
          else
            ENCODED_BLANKS[self.encoding].match?(self)
          end
      end
      x.report("Try Catch Concurrent Map (encode change? #{encoding_required}): ") do
        empty? ||
          begin
            BLANK_RE.match?(self)
          rescue Encoding::CompatibilityError
            ENCODED_BLANKS[self.encoding].match?(self)
          end
      end

    end

    empty? ||
      begin
        BLANK_RE.match?(self)
      rescue Encoding::CompatibilityError
        ENCODED_BLANKS[self.encoding].match?(self)
      end

  end

  private

  def blank_re
    Regexp.new(BLANK_RE.source.encode(self.encoding), Regexp::FIXEDENCODING)
  end
end
```

No matter what there is a price to pay, just from the added code paths. But I think the `Concurrent::Map` caching with a rescue maybe the fastest implementation.

<img width="668" alt="screen shot 2017-11-03 at 4 24 26 pm" src="https://user-images.githubusercontent.com/289882/32394585-fc81598e-c0b3-11e7-8ae2-4dfa8f88ea8d.png">

**Comment by matthewd** [comment]
It looks like there's a real trade-off between pure-ASCII, compatible, and incompatible. The rescue is faster when it's not hit, but much slower when it is.

[My benchmarks](https://gist.github.com/matthewd/7c7e18f14e2247c9f9488e54bed07b76)

Given how rarely-encountered this issue seems to be (most people, most of the time, are using UTF-8), the rescue does seem to be the way to go.

**Comment by kamipo** [comment]
Fixed by #31049.
